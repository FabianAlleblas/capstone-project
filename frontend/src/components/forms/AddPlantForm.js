import styled from 'styled-components/macro'
import useForm from '../../hooks/useForm'
import usePlantList from '../../hooks/usePlantList'
import FormButton from '../buttons/FormButton'

export default function AddPlantForm() {
  const initialInput = {
    id: 20,
    name: '',
    species: '',
  }

  const { saveEntry } = usePlantList()
  const { handleInputChange, handleCancel, formData } = useForm(initialInput)

  function handleSubmit(e) {
    e.preventDefault()
    saveEntry(formData)
    e.target.reset()
  }

  return (
    <Form onSubmit={handleSubmit}>
      <Label>
        Your plants name:
        <Input
          name="name"
          type="text"
          placeholder="Bob"
          maxLength="28"
          onChange={handleInputChange}
        />
      </Label>
      <Label>
        The species of your plant:
        <Input
          name="species"
          type="text"
          placeholder="Monstera deliciosa"
          maxLength="28"
          onChange={handleInputChange}
        />
      </Label>
      <ButtonWrapper>
        <FormButton>Add Plant</FormButton>
        <FormButton secondary onClick={handleCancel}>
          Cancel
        </FormButton>
      </ButtonWrapper>
    </Form>
  )
}

const Form = styled.form`
  align-items: stretch;
  background-color: var(--primary-dark);
  display: grid;
  gap: 24px;
  padding: 0 40px;
`

const Label = styled.label`
  color: var(--form-font-color);
  display: flex;
  flex-wrap: wrap;
  font-size: 1.25rem;
  gap: 8px;
`

const Input = styled.input`
  border-radius: 22px;
  border: none;
  color: var(--form-font-color);
  display: block;
  font-family: 'Josefin sans', sans-serif;
  font-size: 1rem;
  padding: 14px 20px 10px;
  width: 100%;

  &:focus {
    box-shadow: 0 0 4px var(--shadow) inset;
    outline: none;
  }
`

const ButtonWrapper = styled.div`
  display: flex;
  gap: 20px;
`
