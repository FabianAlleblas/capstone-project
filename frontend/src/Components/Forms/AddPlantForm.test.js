import { render, waitFor } from '@testing-library/react'
import user from '@testing-library/user-event'
import AddPlantForm from './AddPlantForm'

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

const onSubmitMock = jest.fn()
const file = new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' })

describe('AddPlantForm', () => {
  it('shows a preview when a image is added to file input', async () => {
    const { getByAltText, getByTestId, getByRole } = render(
      <AddPlantForm savePlantData={onSubmitMock} />
    )

    const imageInput = getByAltText('image-input')
    const ImgInputWrapper = getByTestId('img-input-wrapper')

    user.upload(imageInput, file)

    await waitFor(() => getByRole('button', { name: /ImgDeleteIcon/i }))

    expect(ImgInputWrapper).toHaveStyle(
      'background-image: url(data:image/png;base64,KOKMkOKWoV/ilqEp)'
    )
  })
  it('calls on submit with the correct data, resets the form and calls history push', async () => {
    const { getByLabelText, getByText, getByAltText, getByRole } = render(
      <AddPlantForm savePlantData={onSubmitMock} />
    )

    const imageInput = getByAltText('image-input')

    user.upload(imageInput, file)

    await waitFor(() => getByRole('button', { name: /ImgDeleteIcon/i }))

    user.type(getByLabelText('Your plants name*:'), 'Bob')
    user.type(getByLabelText('The species of your plant*:'), 'Monstera')
    user.type(getByLabelText('Special infos:'), 'variegated')

    user.click(getByText('Add Plant'))

    expect(onSubmitMock).toHaveBeenCalledWith(
      {
        name: 'Bob',
        species: 'Monstera',
        specialInfo: 'variegated',
      },
      {
        name: 'chucknorris.png',
        value: 'data:image/png;base64,KOKMkOKWoV/ilqEp',
      }
    )

    expect(getByLabelText('Your plants name*:')).toHaveValue('')
    expect(getByLabelText('The species of your plant*:')).toHaveValue('')
    expect(getByLabelText('Special infos:')).toHaveValue('')

    expect(mockHistoryPush).toHaveBeenCalled()
  })
  it('calls history push by clicking cancel', () => {
    const { getByText } = render(<AddPlantForm savePlantData={onSubmitMock} />)

    user.click(getByText('Cancel'))

    expect(mockHistoryPush).toHaveBeenCalled()
  })
})
