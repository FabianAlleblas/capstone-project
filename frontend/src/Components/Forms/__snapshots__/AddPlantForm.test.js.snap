// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPlantForm renders correctly 1`] = `
<form
  class="AddPlantForm__Form-sc-11dolld-0 kguMIx"
>
  <label
    class="AddPlantForm__ImgInputWrapper-sc-11dolld-1 iBiABv"
    data-testid="img-input-wrapper"
  >
    <input
      accept="image/png, image/jpeg"
      alt="image-input"
      class="AddPlantForm__ImgInput-sc-11dolld-4 dBfjYL"
      name="image"
      type="file"
    />
    <svg>
      addimgicon.svg
    </svg>
  </label>
  <label
    class="AddPlantForm__Label-sc-11dolld-5 gBBlYq"
  >
    Your plants name*:
    <input
      class="AddPlantForm__Input-sc-11dolld-6 dgePWP"
      maxlength="28"
      name="name"
      placeholder="Bob"
      required=""
      type="text"
    />
  </label>
  <label
    class="AddPlantForm__Label-sc-11dolld-5 gBBlYq"
  >
    The species of your plant*:
    <input
      class="AddPlantForm__Input-sc-11dolld-6 dgePWP"
      maxlength="28"
      name="species"
      placeholder="Monstera deliciosa"
      required=""
      type="text"
    />
  </label>
  <label
    class="AddPlantForm__Label-sc-11dolld-5 gBBlYq"
  >
    Special infos:
    <input
      class="AddPlantForm__Input-sc-11dolld-6 dgePWP"
      maxlength="28"
      name="specialInfo"
      placeholder="variegated"
      type="text"
    />
  </label>
  <div
    class="AddPlantForm__ButtonWrapper-sc-11dolld-7 fQVlqk"
  >
    <button
      class="Button__PrimaryButton-sc-6pjoi6-0 pAbUY AddPlantForm__SubmitButton-sc-11dolld-8 LtjFv"
    >
      Add Plant
    </button>
    <button
      class="Button__SecondaryButton-sc-6pjoi6-1 cSPa-Dk"
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;
